README.MD
Chieru Translator
===

This is a language translator (or encoder, to be more precise) that can translate any sentence to Chieru.

Basic rules are described [here](https://bbs.nga.cn/read.php?tid=21636504).

Android builds are also supported.

CHANGELOG.md
## Unreleased
### Modify
- `fromChieru`: if the "Chieru string" doesn't start with "切噜～♪", return THAT centense rather than "{ERROR}"

src/gui.h
/**
 * @file gui.h
 * @author H1MSK (ksda47832338@outlook.com)
 * @brief Declaration of the main widget
 * @version 0.1
 * @date 2020-07-20
 * 
 * @copyright Copyright (c) 2020
 * 
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the “Software”), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 * 
 * The above copyright notice and this permission notice shall be included in
 * all copies or substantial portions of the Software.
 * 
 * THE SOFTWARE IS PROVIDED “AS IS”, WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
 * THE SOFTWARE.
 */
#pragma once

#include <QWidget>

QT_BEGIN_NAMESPACE
namespace Ui { class EncoderWidget; }
QT_END_NAMESPACE

class ChieruTranslator;
class QTextCodec;
class QPixmap;
class QLabel;

class TranslatorWidget : public QWidget
{
    Q_OBJECT

public:
    TranslatorWidget(QWidget *parent = nullptr);
    ~TranslatorWidget();

    void resizeEvent(QResizeEvent *event) override;

private slots:
    void on_button_from_string_clicked();

    void on_button_to_string_clicked();

    void updateUI();

    void on_combo_encode_currentIndexChanged(const QString &arg1);

private:

    Ui::EncoderWidget *ui;
    ChieruTranslator *m_translator;
    QPixmap *m_horizontal_background, *m_vertical_background, *m_current_background;
    QLabel* m_background_label;
    QTextCodec *m_current_codec;
};


src/cli.cpp
/**
 * @file chieru_translator.cpp
 * @author H1MSK (ksda47832338@outlook.com)
 * @brief Crude cli
 * @version 0.1
 * @date 2020-07-20
 * 
 * @copyright Copyright (c) 2020
 * 
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the “Software”), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 * 
 * The above copyright notice and this permission notice shall be included in
 * all copies or substantial portions of the Software.
 * 
 * THE SOFTWARE IS PROVIDED “AS IS”, WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
 * THE SOFTWARE.
 */

#include <iostream>
#include "chieru_translator.h"

int main() {
    ChieruTranslator *translator = new ChieruTranslator();
    while(1) {
        std::string line;
        std::cout << ">>";
        std::flush(std::cout);
        std::getline(std::cin, line);
        if (line.front() == '1')
            std::cout <<translator->toUTF8(line.substr(2)) <<std::endl;
        else if (line.front() == '0')
            std::cout <<translator->fromUTF8(line.substr(2)) <<std::endl;
        else if (line.front() == 'q')
            exit(0);
    }
}

src/singleton.h
/**
 * @file singleton.h
 * @author H1MSK (ksda47832338@outlook.com)
 * @brief A handy singleton library
 * @version 0.3
 * @date 2020-07-20
 * 
 * @copyright Copyright (c) 2020
 * 
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the “Software”), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 * 
 * The above copyright notice and this permission notice shall be included in
 * all copies or substantial portions of the Software.
 * 
 * THE SOFTWARE IS PROVIDED “AS IS”, WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
 * THE SOFTWARE.
 */

#include <cassert>
#include <mutex>

/**
 * @brief A handy macro that defines a private constructor and
 *          give access to the corresponding singleton function
 * 
 * @warning This macro has a private access modifier inside
 */
#define SINGLETON_CONTRUCTOR(class_name, ...) \
    friend class_name* Instance(__VA_ARGS__); \
 private: \
    class_name(__VA_ARGS__)

/**
 * @brief Wrapper class for the instance to support contruct-on-need
 * 
 * @tparam Type The class type
 */
template <typename Type>
struct InstanceWrapper {
    bool initialized;
    Type* raw_pointer;

    Type* operator->() {
        assert(initialized == true);
        return raw_pointer;
    }
    const Type* operator->() const {
        assert(initialized = true);
        return raw_pointer;
    }

    operator Type*() {
        assert(initialized = true);
        return raw_pointer;
    }
};

/**
 * @brief Class that actually stores the instance and the access mutex
 * 
 * @tparam Type The type of the class
 */
template <typename Type>
struct InstanceSafetyHelper {
    InstanceWrapper<Type> wrapper;
    std::mutex mutex;

    static InstanceSafetyHelper* Helper() {
        static InstanceSafetyHelper<Type> helper;
        return &helper;
    }
    InstanceSafetyHelper() : wrapper{false, nullptr}, mutex() {}
    ~InstanceSafetyHelper() {
        wrapper.raw_pointer->~Type();
        free(wrapper.raw_pointer);
    }
};

/**
 * @brief Get the instance of the indicate type, construct one when not
 * constructed (thread safe)
 *
 * @tparam Type The type of the instance
 * @tparam ConstructParameters Parameters used in construction
 * @param args Parameters used in construction
 * @return Type* The instance
 */
template <typename Type, typename... ConstructParameters>
Type* Instance(ConstructParameters... args) {
    InstanceSafetyHelper<Type>* helper = InstanceSafetyHelper<Type>::Helper();
    if (helper->wrapper.raw_pointer == nullptr) {
        helper->mutex.lock();
        if (helper->wrapper.raw_pointer == nullptr) {
            void* data = malloc(sizeof(Type));
            helper->wrapper.raw_pointer =
                reinterpret_cast<Type*>(data);
            helper->wrapper.raw_pointer = new (data) Type(args...);
            helper->wrapper.initialized = true;
        }
        helper->mutex.unlock();
    }
    return helper->wrapper;
}
/**
 * @brief Get the Instance with the indicated type after asserting its existance
 *
 * @tparam Type The type of the instance
 * @return InstanceWrapper<Type>* The instance
 */
template <typename Type>
InstanceWrapper<Type> getInstance(void) {
    InstanceSafetyHelper<Type>* helper = InstanceSafetyHelper<Type>::Helper();
    assert(helper->wrapper.raw_pointer != nullptr);
    return helper->wrapper;
}


src/chieru_translator.h
/**
 * @file chieru_translator.h
 * @author H1MSK (ksda47832338@outlook.com)
 * @brief Declaration of Chieru translator
 * @version 0.1
 * @date 2020-07-18
 *
 * @warning This file should be encoded in UTF-8
 *
 * @copyright Copyright (c) 2020
 *
 */

#include <QString>
#include <QHash>
#include <QTextCodec>

class ChieruTranslator {
 private:
    static bool s_initialized;
    static QChar s_chieru_charactor[16];
    static QHash<QChar, int> s_dict;
    static QString s_symbols;


 protected:
    static QString word2chieru(QByteArray::const_iterator begin,
                               QByteArray::const_iterator end);
    static QByteArray chieru2word(QString::const_iterator begin,
                                  QString::const_iterator end);

    static bool is_separator(QChar ch) {
        if ((ch < 0x30) ||                      // before '0'
            (ch > 0x39 && ch <= 0x40) ||        // between '9' and 'A'
            (ch > 0x5A && ch <= 0x60) ||        // between 'Z' and 'a'
            (ch > 0x7A && ch <= 0x7F))          // after 'z' in ascii
            return true;

        // is Chinese symbol
        for (QChar symbol : s_symbols)
            if (ch == symbol) return true;

        return false;
    }

 public:
    ChieruTranslator();

    void initialize();
    static QString word2chieru(const QByteArray& word);
    static QByteArray chieru2word(const QString& word);
    QString toChieru(const QString& string, QTextCodec* codec = QTextCodec::codecForName("UTF8"));
    QString fromChieru(const QString& string, QTextCodec* codec = QTextCodec::codecForName("UTF8"));
};


src/abstract_translator.h
/**
 * @file abstract_translator.h
 * @author H1MSK (ksda47832338@outlook.com)
 * @brief Abstract base translator class
 * @version 0.1
 * @date 2020-07-20
 * 
 * @copyright Copyright (c) 2020
 * 
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the “Software”), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 * 
 * The above copyright notice and this permission notice shall be included in
 * all copies or substantial portions of the Software.
 * 
 * THE SOFTWARE IS PROVIDED “AS IS”, WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
 * THE SOFTWARE.
 */

#pragma once

#include <string>
#include <map>
#include <any>

class AbstractTranslator {
public:
    AbstractTranslator() { initialize(); }
    virtual ~AbstractTranslator() {}

    virtual void initialize() { };

    virtual std::string fromUTF8(const std::string& string) = 0;
    virtual std::string toUTF8(const std::string& string) = 0;

#ifdef QT
    virtual bool hasConfigWidget() { return false; }
    virtual void openConfigWidget() { }
#endif
};


src/gui.cpp
/**
 * @file gui.cpp
 * @author H1MSK (ksda47832338@outlook.com)
 * @brief GUI wrapper for the translator
 * @version 0.1
 * @date 2020-07-20
 * 
 * @copyright Copyright (c) 2020
 * 
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the “Software”), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 * 
 * The above copyright notice and this permission notice shall be included in
 * all copies or substantial portions of the Software.
 * 
 * THE SOFTWARE IS PROVIDED “AS IS”, WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
 * THE SOFTWARE.
 */

#include "gui.h"
#include "ui_gui.h"

#include <QTextCodec>
#include <QResizeEvent>
#include <QDateTime>
#include <QPixmap>
#include <QLabel>
#include <QGraphicsOpacityEffect>
#include <QDebug>
#include <QFontDatabase>

#include <cstdlib>

#include "chieru_translator.h"
#include "singleton.h"

TranslatorWidget::TranslatorWidget(QWidget *parent)
    : QWidget(parent)
    , ui(new Ui::EncoderWidget)
    , m_translator(Instance<ChieruTranslator>())
    , m_horizontal_background(nullptr)
    , m_vertical_background(nullptr)
    , m_current_background(nullptr)
    , m_background_label(new QLabel(this))
    , m_current_codec(QTextCodec::codecForName("UTF8")) {
    ui->setupUi(this);

    srand(QDateTime::currentMSecsSinceEpoch());
    int total;
    for (total = 0; QFile::exists(QString(":/imgs/chieruh%1.png").arg(total)); ++total) {
        qDebug() <<"Found: " <<QString(":/imgs/chieruh%1.png").arg(total);
    }
    m_horizontal_background = new QPixmap(QString(":/imgs/chieruh%1.png").arg(rand() % total));

    for (total = 0; QFile::exists(QString(":/imgs/chieruv%1.png").arg(total)); ++total) {
        qDebug() <<"Found: " <<QString(":/imgs/chieruv%1.png").arg(total);
    }
    m_vertical_background = new QPixmap(QString(":/imgs/chieruv%1.png").arg(rand() % total));

    m_current_background = m_horizontal_background;

    m_background_label->setAttribute(Qt::WA_TransparentForMouseEvents);
    m_background_label->resize(size());
    m_background_label->show();
    m_background_label->setPixmap(*m_horizontal_background);

    QGraphicsOpacityEffect *effect = new QGraphicsOpacityEffect(m_background_label);
    effect->setEnabled(true);
    effect->setOpacity(0.2);
    m_background_label->setGraphicsEffect(effect);

#ifdef ANDROID
    QFont f = ui->textedit_original->font();
    f.setPointSize(18);
    ui->textedit_original->setFont(f);
    ui->textedit_translated->setFont(f);
#endif

    ui->combo_chieru->hide();

    updateUI();
}

TranslatorWidget::~TranslatorWidget()
{
    delete ui;
}

void TranslatorWidget::resizeEvent(QResizeEvent *event) {
    m_background_label->resize(event->size());
    updateUI();
}

void TranslatorWidget::updateUI() {
    do {
        if ((ui->central_layout->direction() == QBoxLayout::LeftToRight)) {
            if (width() > height()) break;

            // Change direction to vertical
            ui->central_layout->setDirection(QBoxLayout::TopToBottom);
            ui->layout_control->setDirection(QBoxLayout::LeftToRight);
            ui->button_to_string->setText(QString("\xE2\x86\x91")); // ↑
            ui->button_from_string->setText(QString("\xE2\x86\x93")); // ↓

            m_current_background = m_vertical_background;
        } else {
            if (width() < height()) break;

            // Change direction to horizontal
            ui->central_layout->setDirection(QBoxLayout::LeftToRight);
            ui->layout_control->setDirection(QBoxLayout::TopToBottom);
            ui->button_to_string->setText(QString("\xE2\x86\x90")); // ←
            ui->button_from_string->setText(QString("\xE2\x86\x92")); // →

            m_current_background = m_horizontal_background;
        }
    } while(0);

    // Change background and center it
    QPixmap background = m_current_background->scaled(size(), Qt::KeepAspectRatioByExpanding);

    m_background_label->resize(background.size());
    if (background.width() > width()) {
        m_background_label->move((width() - background.width()) / 2, 0);
    } else {
        m_background_label->move(0, (height() - background.height()) / 2);
    }
    m_background_label->setPixmap(background);

    m_background_label->raise();
}

void TranslatorWidget::on_button_from_string_clicked() {
    QString translated_string =
            m_translator->toChieru(ui->textedit_original->toPlainText(), m_current_codec);

    ui->textedit_translated->setPlainText(translated_string);
}

void TranslatorWidget::on_button_to_string_clicked() {
    QString origianl_string =
            m_translator->fromChieru(ui->textedit_translated->toPlainText(), m_current_codec);

    ui->textedit_original->setPlainText(origianl_string);
}

void TranslatorWidget::on_combo_encode_currentIndexChanged(const QString &arg1) {
    m_current_codec = QTextCodec::codecForName(arg1.toUtf8());
}

int main(int argc, char *argv[]) {
    QApplication a(argc, argv);

    int font_id = QFontDatabase::addApplicationFont(":/fonts/FeiGunGun.ttf");
    if (font_id != -1) {
        QFont font(QFontDatabase::applicationFontFamilies(font_id).at(0));
        QGuiApplication::setFont(font);
    }

    Instance<TranslatorWidget>()->show();
    return a.exec();
}
